https://issues.redhat.com/browse/ACM-16419 Replace custom metrics doc (#7501)
open-cluster-management/rhacm-docs: observability/customize_observability.adoc

* Find the metric type by checking the annotation that the `ServiceMonitor` for the observed resource uses.
** Platform metrics use `operator.prometheus.io/controller-id: openshift-platform-monitoring/prometheus-operator`.
** User workload metrics use `operator.prometheus.io/controller-id: openshift-user-workload-monitoring/prometheus-operator`.

You can also find `ServiceMonitors` in the console by going to *Observe* > *Targets* and choosing Platform or User from the *Source* filter in the top right. 

*Note:* The *Source* filter provides service monitor or target information, not a list of metrics.

If the metric is a Platform, continue at <<adding-platform-metrics,Adding Platform metrics>>. If the metric is a User workload, continue at <<adding-user-workload-metrics,Adding user workload metrics>>.

[#adding-platform-metrics]
=== Adding Platform metrics

You can monitor Platform metrics by creating a `ConfigMap` on the hub cluster in the `open-cluster-management-observability` namespace. Use `observability-metrics-custom-allowlist` as the name. See the following `ConfigMap` example that you can use to monitor Platform metrics:

[source,yaml]
----
kind: ConfigMap
apiVersion: v1

> metadata:

Reviewer: Format?

name: observability-metrics-custom-allowlist
namespace: open-cluster-management-observability
data:
metrics_list.yaml: |
names: <1>
  - node_memory_MemTotal_bytes

>     recording_ules: <2>

Reviewer 2: ```suggestion
    recording_rules: <2>
```

  - record: apiserver_request_duration_seconds:histogram_quantile_90
    expr: histogram_quantile(0.90,sum(rate(apiserver_request_duration_seconds_bucket {job=\"apiserver\", verb!=\"WATCH\"}[5m])) by (verb,le))
----

<1> *Optional:* Add the names of the custom metrics you want to collect from your managed cluster.
<2> *Optional:* Add one value for the `expr` and `record` parameter pair to define the query expression.

The name for the metrics you collect is the same name that you define in your managed cluster `record` parameter. After you run the query expression, you get the metric value results. You can use one section or both sections, which applies to every cluster with monitoring enabled.

If you only want to collect custom metrics from a single managed cluster, use the following example and apply the config map on your managed cluster in the `open-cluster-management-addon-observability` namespace:

[source,yaml]
----
kind: ConfigMap
apiVersion: v1
metadata:
name: observability-metrics-custom-allowlist
namespace: open-cluster-management-addon-observability
data:
metrics_list.yaml: |
names: <1>
  - node_memory_MemTotal_bytes

>     rules: <2>

Reviewer 2: ```suggestion
    recording_rules: <2>
```

  - record: apiserver_request_duration_seconds:histogram_quantile_90
    expr: histogram_quantile(0.90,sum(rate(apiserver_request_duration_seconds_bucket{job="apiserver", verb!="WATCH"}[5m])) by (verb,le))
----

<1> *Optional:* Add the names of the custom metrics you want to collect from your managed cluster.
<2> Optional: Enter only one value for the `expr` and `record` parameter pair to define the query expression.

The name for the metrics you collect is the same name that you define in your managed cluster `record` parameter. After you run the query expression, you get the metric value results. You can use one section or both sections.

[#adding-user-workload-metrics]
=== Adding User workload metrics

You can monitor User workload metrics by setting the configuration on the managed cluster in namespace where you want to capture the metric. The name must be `observability-metrics-custom-allowlist` and the format must be the same as the following example:

[source,yaml]
----
kind: ConfigMap
apiVersion: v1
metadata:

> name: observability-metrics-custom-allowlist

Reviewer: This format looks way off, there is usually the indention under metadata. See the repo for how this is done, even with name.

Example:

```
[source,yaml]
----
apiVersion: cluster.open-cluster-management.io/v1beta1
kind: ClusterCurator
metadata:
  name: my-cluster
  namespace: clusters
  labels:
    open-cluster-management: curator
```

[#adding-custom-metrics]
== Adding custom metrics

To monitor metrics from a remote cluster by using {acm}, check if the metric is exported as a Platform or a User workload metric. Use one of the following four methods to find the metric type:

* Find the metric type in the documentation of the solution you want to monitor.

* Find the metric type by contacting the support for your product.

> * Find the metric type in the console by selecting your cluster from the dropdown and navigating to *Observe* > *Metrics*.

Reviewer 2: I am not fully understanding exactly these steps. I looked in a console with user workload metrics enabled, but with no targets. I saw no metrics annotated as "Platform" anywhere.

Author: @gerikis

** In the Prometheus column, Platform metrics are listed as the metric provider.
** In the Prometheus column, User workload metrics are listed as `openshift-user-workload-monitoring`.

> * Find the metric type by checking the annotation that the `ServiceMonitor` for the observed resource uses.

Reviewer 2: You can also find ServiceMonitors in the webconsole by going to observe->targets. On this page there is a filter on the top right corner where one can choose the "Source" Platform or User. However this doesn't give an actual list of metrics, just the service monitor/target information

Author: Is this something you think we should add? @jacobbaungard @gerikis? I would probably add it as a note if we want to add it.

Reviewer 2: I think that might work well as an additional note as suggested, but not essential.

Author: Added, thanks!

** Platform metrics use `operator.prometheus.io/controller-id: openshift-platform-monitoring/prometheus-operator`.
** User workload metrics use `operator.prometheus.io/controller-id: openshift-user-workload-monitoring/prometheus-operator`.

If the metric is a Platform, continue at <<adding-platform-metrics,Adding Platform metrics>>. If the metric is a User workload, continue at <<adding-user-workload-metrics,Adding user workload metrics>>.

[#adding-platform-metrics]
=== Adding Platform metrics

You can monitor Platform metrics by creating a `ConfigMap` on the hub cluster in the `open-cluster-management-observability` namespace. Use `observability-metrics-custom-allowlist` as the name. See the following `ConfigMap` example that you can use to monitor Platform metrics:

[source,yaml]
----
kind: ConfigMap
apiVersion: v1
metadata:
name: observability-metrics-custom-allowlist
namespace: open-cluster-management-observability
data:
metrics_list.yaml: |
names: <1>
  - node_memory_MemTotal_bytes
    rules: <2>

>   - record: apiserver_request_duration_seconds:histogram_quantile_90

Reviewer 2: Did we have this example from before? I believe `rules` are alerting rules and would normally need something like `>0` (i.e there needs to be a condition that can evalutate to true/false).

The example here is a recording rule, and I'm pretty sure this needs to go under `recording_rules` and not just `rules`. Adding alerting rules here is probably not the best ideas in most cases.

Will need to double check this, but pretty sure it needs to be like this.

Author: Any update here @jacobbaungard? I can add `recording_rules` if it's missing and is required.

Reviewer 2: Confirmed, please use `recording_rules` instead of `rules`.

Author: Updated

    expr: histogram_quantile(0.90,sum(rate(apiserver_request_duration_seconds_bucket {job=\"apiserver\", verb!=\"WATCH\"}[5m])) by (verb,le))
----

<1> *Optional:* Add the names of the custom metrics you want to collect from your managed cluster.
<2> *Optional:* Add one value for the `expr` and `record` parameter pair to define the query expression.

The name for the metrics you collect is the same name that you define in your managed cluster `record` parameter. After you run the query expression, you get the metric value results. You can use one section or both sections, which applies to every cluster with monitoring enabled.

> If you want to use the configuration for a single cluster, use the following example and on your managed cluster in the `open-cluster-management-addon-monitoring` namespace:

Author: Please check this, the paragraph didn't make much sense to me in the original draft.

Reviewer 2: The example above which creates a configmap on the Hub, will collect those metrics for all managed clusters.

This section, is about collecting custom metrics from just a single managed-cluster.

```suggestion
<2> *Optional:* Add one value for the `expr` and `record` parameter pair to define the query expression.

The name for the metrics you collect is the same name that you define in your managed cluster `record` parameter. After you run the query expression, you get the metric value results. You can use one section or both sections, which applies to every cluster with monitoring enabled.

If you only want to collect custom metrics from a single managed cluster, use the following example and apply the config map on your managed cluster in the `open-cluster-management-addon-observability` namespace:
```

Author: Updated

    expr: histogram_quantile(0.90,sum(rate(apiserver_request_duration_seconds_bucket{job="apiserver", verb!="WATCH"}[5m])) by (verb,le))
----

<1> *Optional:* Add the names of the custom metrics you want to collect from your managed cluster.
<2> Optional: Enter only one value for the `expr` and `record` parameter pair to define the query expression.

The name for the metrics you collect is the same name that you define in your managed cluster `record` parameter. After you run the query expression, you get the metric value results. You can use one section or both sections.

[#adding-user-workload-metrics]
=== Adding User workload metrics

You can monitor User workload metrics by setting the configuration on the managed cluster in namespace where you want to capture the metric. The name must be `observability-metrics-custom-allowlist` and the format must be the same as the following example:

[source,yaml]
----
kind: ConfigMap
apiVersion: v1
metadata:
name: observability-metrics-custom-allowlist
namespace: <monitored_namespace> <1>
data:
uwl_metrics_list.yaml: <2>
names:
  <sample_metrics> <3>
----

<1> Add the namespace where you want to capture the metric.

> <2> Add the key for the config map data.

Reviewer 2: This key is not user configurable and always needs to be `uwl_metrics_list.yaml`, so I think we could just remove the additional annotation (i.e the <2> there).

Author: Updated

<3> Add the value of the config map data in YAML format. Add list of metric names that you want to collect from the `test` namespace to the `names` section. After you create the config map, the observability collector collects and pushes the metrics from the target namespace to the hub cluster.

> The previous example monitors the User workload metric `sample_metrics` from the namespace `monitored_namespace`. If you create the configuration in the `open-cluster-management-addon-monitoring` namespace instead, the metrics are collected from all the namespaces of the managed cluster.

Reviewer 2: ```suggestion
The previous example monitors the User workload metric `sample_metrics` from the namespace `monitored_namespace`. If you create the configuration in the `open-cluster-management-addon-observability` namespace instead, the metrics are collected from all the namespaces of the managed cluster.
```
